lua/weapons/weapon_ttt_minty_blink/shared.lua

```diff
@@ -0,0 +1,471 @@
+SWEP.ViewModelFOV           = 67
+SWEP.WorldModel             = "models/weapons/w_slam.mdl"
+
+SWEP.Primary.Recoil         = 0
+SWEP.Primary.ClipSize       = -1
+SWEP.Primary.DefaultClip    = -1
+SWEP.Primary.Automatic      = false
+SWEP.Primary.Delay          = 1
+SWEP.Primary.Ammo           = "none"
+
+--  TTT
+SWEP.CanBuy                 = { ROLE_TRAITOR }
+SWEP.LimitedStock           = true
+SWEP.NoSights               = true
+
+SWEP.EquipMenuData          = {
+    type = "item_weapon",
+    desc = "Woosh woosh.\n\nPrimary: Hold to target\nSecondary: Abort blink"
+};
+
+SWEP.Icon                   = res.icon
+
+--  Special
+SWEP.Timer      = {
+    Particle    = 0,
+    Recharge    = 0,
+    PostProcess = 0,
+}
+
+SWEP.Charge = {
+    Current = ( cvar.charge_count:GetInt() > 0 and cvar.charge_count:GetInt() or cvar.charge_max:GetInt() ),
+    Maximum = ( cvar.charge_count:GetInt() > 0 and cvar.charge_count:GetInt() or cvar.charge_max:GetInt() ),
+}
+
+SWEP.Warp   = {
+    Marker  = nil,
+    Target  = nil,
+    Player  = nil,
+}
+
+SWEP.Emitter    = nil
+SWEP.State      = nil
+
+-- Logic
+
+--  Default
+function SWEP:DryFire() return false end        -- Disable default behaviours
+function SWEP:PrimaryAttack() return false end
+
+function SWEP:Initialize()
+    self:SetHoldType( self.HoldType )
+    
+    local charge_max = ( cvar.charge_count:GetInt() > 0 and cvar.charge_count:GetInt() or cvar.charge_max:GetInt() )
+    self.Charge.Current = charge_max
+    self.Charge.Maximum = charge_max
+
+    if CLIENT then
+        self.Emitter = ParticleEmitter( self:GetPos() )
+    end
+
+    self.BaseClass.Initialize( self )
+end
+
+function SWEP:OwnerChanged()
+    self:Blink_ClampCharge()
+    self:Blink_ForceReset()
+
+    self.BaseClass.OwnerChanged( self )
+end
+
+function SWEP:OnRemove()
+    self:Blink_SetTimers()
+    if self.Emitter then self.Emitter:Finish() end
+end
+
+function SWEP:PreDrawViewModel()
+    render.SetBlend( 0 ) -- Hide viewmodel
+end
+
+function SWEP:ViewModelDrawn()                  
+    render.SetBlend( 1 ) -- Show hands
+end
+
+function SWEP:Deploy()
+    self:Blink_DoViewModelAnimation()
+    return true
+end
+
+--  Utility
+function SWEP:Blink_SetTimers( particle, recharge, post_process )
+    self.Timer.Particle = particle and CurTime() or 0
+    self.Timer.Recharge = recharge and CurTime() or 0
+    self.Timer.PostProcess = post_process and CurTime() or 0
+end
+
+function SWEP:Blink_DoViewModelAnimation()
+    local owner = self.Owner
+    if not IsValid( owner ) then return end
+
+    local viewmodel = owner:GetViewModel()
+    if not IsValid( viewmodel ) then return end
+
+    if SERVER or IsFirstTimePredicted() then
+        viewmodel:SendViewModelMatchingSequence( 12 )
+    end
+end
+
+function SWEP:Blink_DoAimTrace()
+    local owner = self.Owner
+    if not IsValid( owner ) then return nil end
+
+    local trace_hull = Vector( trace_hull_size, trace_hull_size, trace_hull_size )
```

Comment:
`trace_hull` could be created once outside of this function, no need to create the same Vector over and over